src/lib/comms/sol-mqtt.c

```diff
@@ -0,0 +1,579 @@
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <mosquitto.h>
+#include <stdbool.h>
+
+#define SOL_LOG_DOMAIN &_sol_mqtt_log_domain
+#include <sol-log-internal.h>
+#include <sol-mainloop.h>
+#include <sol-util.h>
+
+#include "sol-mqtt.h"
+
+SOL_LOG_INTERNAL_DECLARE(_sol_mqtt_log_domain, "mqtt");
+
+static int init_ref;
+#define CHECK_INIT(ret) \
+    do { \
+        if (unlikely(init_ref < 1)) { \
+            SOL_WRN("sol-mqtt used before initialization"); \
+            return ret; \
+        } \
+    } while (0)
+
+#define MQTT_CHECK_API(ptr, ...) \
+    do {                                        \
+        if (unlikely(ptr->api_version != \
+            SOL_MQTT_CONFIG_API_VERSION)) { \
+            SOL_WRN("Couldn't handle mqtt handler that has unsupported " \
+                "version '%u', expected version is '%u'", \
+                ptr->api_version, SOL_MQTT_CONFIG_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0)
+
+struct sol_mqtt {
+    struct mosquitto *mosq;
+
+    struct sol_fd *socket_read;
+    struct sol_fd *socket_write;
+
+    struct sol_timeout *connect_timeout;
+    struct sol_timeout *disconnect_timeout;
+    struct sol_timeout *publish_timeout;
+    struct sol_timeout *subscribe_timeout;
+    struct sol_timeout *unsubscribe_timeout;
+    struct sol_timeout *message_timeout;
+
+    const struct sol_mqtt_config *config;
```

Comment:
if we're keeping the handle, we must document that, so users don't allocate this on a stack that may go away.

IMO we should keep a copy (ie: inline `struct sol_mqtt_config` and `memcpy()` it).